In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
## obtain the datasets for first two stocks given
dfA1=pd.read_csv(r"sp500_tickers_A-D_1min_1pppix\A_1min.txt",header=None)
dfAA1=pd.read_csv(r"sp500_tickers_A-D_1min_1pppix\AA_1min.txt",header=None)

In [3]:
## define the columns of both the dataframes
dfA1.columns=["date","Opening Price","High","Low","Closing Price","Volume"]
dfAA1.columns=["date","Opening Price","High","Low","Closing Price","Volume"]

In [4]:
dfA1

date  Opening Price      High       Low  \
0        2005-01-03 09:30:00        14.8242   14.8242   14.8181   
1        2005-01-03 09:31:00        14.8242   14.8242   14.8119   
2        2005-01-03 09:32:00        14.8119   14.8242   14.8119   
3        2005-01-03 09:33:00        14.8181   14.8242   14.8181   
4        2005-01-03 09:34:00        14.8242   14.8242   14.8181   
...                      ...            ...       ...       ...   
1680045  2022-02-25 16:02:00       133.7300  133.7300  133.7300   
1680046  2022-02-25 16:03:00       133.7300  133.7300  133.7300   
1680047  2022-02-25 16:10:00       133.7300  133.7300  133.7300   
1680048  2022-02-25 18:30:00       133.7300  133.7300  133.7300   
1680049  2022-02-25 19:00:00       133.7300  133.7300  133.7300   

         Closing Price  Volume  
0              14.8242  173492  
1              14.8119    9646  
2              14.8181   34810  
3              14.8242   11464  
4              14.8181   12302  
...                ...     ...  
1680045       133.7300  354729  
1680046       133.7300    6777  
1680047       133.7300       0  
1680048       133.7300       0  
1680049       133.7300       0  

[1680050 rows x 6 columns]

In [5]:
## split the date column as it contains the date and time both values of the stock prices present in the dataset
dfA1[['Date', 'Time']] = dfA1['date'].str.split(' ', 1, expand=True)
## split the date column as it contains the date and time both values of the stock prices present in the dataset
dfAA1[['Date', 'Time']] = dfAA1['date'].str.split(' ', 1, expand=True)

C:\Users\kumar\AppData\Local\Temp\ipykernel_18140\1298526840.py:2: FutureWarning: In a future version of pandas all arguments of StringMethods.split except for the argument 'pat' will be keyword-only.
  dfA1[['Date', 'Time']] = dfA1['date'].str.split(' ', 1, expand=True)
C:\Users\kumar\AppData\Local\Temp\ipykernel_18140\1298526840.py:4: FutureWarning: In a future version of pandas all arguments of StringMethods.split except for the argument 'pat' will be keyword-only.
  dfAA1[['Date', 'Time']] = dfAA1['date'].str.split(' ', 1, expand=True)


In [6]:
dfA1

date  Opening Price      High       Low  \
0        2005-01-03 09:30:00        14.8242   14.8242   14.8181   
1        2005-01-03 09:31:00        14.8242   14.8242   14.8119   
2        2005-01-03 09:32:00        14.8119   14.8242   14.8119   
3        2005-01-03 09:33:00        14.8181   14.8242   14.8181   
4        2005-01-03 09:34:00        14.8242   14.8242   14.8181   
...                      ...            ...       ...       ...   
1680045  2022-02-25 16:02:00       133.7300  133.7300  133.7300   
1680046  2022-02-25 16:03:00       133.7300  133.7300  133.7300   
1680047  2022-02-25 16:10:00       133.7300  133.7300  133.7300   
1680048  2022-02-25 18:30:00       133.7300  133.7300  133.7300   
1680049  2022-02-25 19:00:00       133.7300  133.7300  133.7300   

         Closing Price  Volume        Date      Time  
0              14.8242  173492  2005-01-03  09:30:00  
1              14.8119    9646  2005-01-03  09:31:00  
2              14.8181   34810  2005-01-03  09:32:00  
3              14.8242   11464  2005-01-03  09:33:00  
4              14.8181   12302  2005-01-03  09:34:00  
...                ...     ...         ...       ...  
1680045       133.7300  354729  2022-02-25  16:02:00  
1680046       133.7300    6777  2022-02-25  16:03:00  
1680047       133.7300       0  2022-02-25  16:10:00  
1680048       133.7300       0  2022-02-25  18:30:00  
1680049       133.7300       0  2022-02-25  19:00:00  

[1680050 rows x 8 columns]

In [7]:
## remove the date column now as we have extracted it into two different columns
dfA1=dfA1.iloc[:,1:]
dfAA1=dfAA1.iloc[:,1:]
dfAA1

Opening Price     High      Low  Closing Price  Volume        Date  \
0             26.5376  26.5376  26.5376        26.5376     100  2016-10-18   
1             26.5873  26.5972  26.5873        26.5972     400  2016-10-18   
2             26.5972  26.5972  26.5873        26.5873     700  2016-10-18   
3             26.5773  26.5773  26.5376        26.5376     700  2016-10-18   
4             26.5376  26.5376  26.5376        26.5376     500  2016-10-18   
...               ...      ...      ...            ...     ...         ...   
575054        77.7800  77.7800  77.7800        77.7800     775  2022-02-25   
575055        77.7700  77.7700  77.7700        77.7700     100  2022-02-25   
575056        77.7700  77.7700  77.7700        77.7700     369  2022-02-25   
575057        77.7900  77.7900  77.7900        77.7900    1175  2022-02-25   
575058        77.9000  77.9000  77.7900        77.7900    1000  2022-02-25   

            Time  
0       04:04:00  
1       04:10:00  
2       04:11:00  
3       04:14:00  
4       06:30:00  
...          ...  
575054  19:10:00  
575055  19:23:00  
575056  19:45:00  
575057  19:47:00  
575058  19:59:00  

[575059 rows x 7 columns]

#### Q 1(a)

In [8]:
## to plot the minute by minute closing prices
## group the first dataset by date
group=dfA1.groupby("Date")
## choose a particular date and plot the values
dfA1day1=group.get_group("2005-01-03")
## group the second dataset by date
group=dfAA1.groupby("Date")
## choose a particular date and plot the values
dfAA1day1=group.get_group("2016-10-18")
dfAA1day1

Opening Price     High      Low  Closing Price  Volume        Date  \
0          26.5376  26.5376  26.5376        26.5376     100  2016-10-18   
1          26.5873  26.5972  26.5873        26.5972     400  2016-10-18   
2          26.5972  26.5972  26.5873        26.5873     700  2016-10-18   
3          26.5773  26.5773  26.5376        26.5376     700  2016-10-18   
4          26.5376  26.5376  26.5376        26.5376     500  2016-10-18   
..             ...      ...      ...            ...     ...         ...   
443        26.2891  26.2891  26.2891        26.2891    1000  2016-10-18   
444        26.2891  26.2891  26.2891        26.2891     400  2016-10-18   
445        26.3388  26.3388  26.3388        26.3388    1000  2016-10-18   
446        26.2891  26.2891  26.2891        26.2891    1000  2016-10-18   
447        26.2394  26.2394  26.2394        26.2394    1000  2016-10-18   

         Time  
0    04:04:00  
1    04:10:00  
2    04:11:00  
3    04:14:00  
4    06:30:00  
..        ...  
443  17:48:00  
444  17:50:00  
445  19:24:00  
446  19:32:00  
447  19:59:00  

[448 rows x 7 columns]

In [9]:
import plotly.graph_objects as go
## Plot the closing prices with time in the x axis and price ini the y axis
fig = go.Figure(data=go.Scatter(x=dfA1day1["Time"], y=dfA1day1["Closing Price"], mode='lines'))
## Plot the closing prices with time in the x axis and price ini the y axis
fig.update_layout(
    title='Closing Price',
    xaxis=dict(title='Time'),
    yaxis=dict(title='Price')
)

# Show the plot
fig.show()


In [10]:
import plotly.graph_objects as go
## Plot the closing prices with time in the x axis and price ini the y axis
fig = go.Figure(data=go.Scatter(x=dfAA1day1["Time"], y=dfAA1day1["Closing Price"], mode='lines'))
## Plot the closing prices with time in the x axis and price ini the y axis

fig.update_layout(
    title='Closing Price',
    xaxis=dict(title='Time'),
    yaxis=dict(title='Price')
)

# Show the plot
fig.show()


### Q 1(b)

In [11]:
### get day by day data for a stock
def getdata(df):
    ## group the dataset by date and take the last index to get the closing proce for each day
    df1=df.groupby("Date").last().reset_index()
    df1=df1.head(300)
    return df1

In [12]:
## get the day by day data for a particular stock
dfA1daybyday=getdata(dfA1)
dfA1daybyday

Date  Opening Price     High      Low  Closing Price  Volume  \
0    2005-01-03        14.4551  14.4551  14.4551        14.4551     140   
1    2005-01-04        14.3383  14.3383  14.3014        14.3075   21110   
2    2005-01-05        14.2952  14.2952  14.2952        14.2952     559   
3    2005-01-06        13.9877  13.9877  13.9877        13.9877     140   
4    2005-01-07        13.9612  13.9612  13.9612        13.9612     280   
..          ...            ...      ...      ...            ...     ...   
295  2006-03-07        22.3593  22.3593  22.3593        22.3593     559   
296  2006-03-08        22.4762  22.4885  22.4762        22.4824   10485   
297  2006-03-09        22.2732  22.2732  22.2732        22.2732    1538   
298  2006-03-10        22.3839  22.3962  22.3778        22.3901   10345   
299  2006-03-13        22.7161  22.7161  22.7161        22.7161   62211   

         Time  
0    19:25:00  
1    15:59:00  
2    17:22:00  
3    18:51:00  
4    17:27:00  
..        ...  
295  16:47:00  
296  15:59:00  
297  16:03:00  
298  15:59:00  
299  16:47:00  

[300 rows x 7 columns]

In [13]:
fig = go.Figure(data=go.Scatter(x=dfA1daybyday["Date"], y=dfA1daybyday["Closing Price"], mode='lines'))

## get the day by day data for a particular stock
fig.update_layout(
    title='Closing Price',
    xaxis=dict(title='Time'),
    yaxis=dict(title='Price')
)

# Show the plot
fig.show()


In [14]:
## similarly do it for the other stock
dfAA1daybyday=getdata(dfAA1)
dfAA1daybyday

Date  Opening Price     High      Low  Closing Price  Volume  \
0    2016-10-18        26.2394  26.2394  26.2394        26.2394    1000   
1    2016-10-19        26.8954  26.8954  26.8854        26.8854     500   
2    2016-10-20        26.6171  26.6171  26.6171        26.6171     136   
3    2016-10-21        26.7662  26.7662  26.7662        26.7662     100   
4    2016-10-24        26.7562  26.7562  26.7562        26.7562     100   
..          ...            ...      ...      ...            ...     ...   
295  2017-12-19        46.4985  46.4985  46.4985        46.4985     100   
296  2017-12-20        49.2924  49.2924  49.2924        49.2924     100   
297  2017-12-21        49.0429  49.0429  49.0429        49.0429     100   
298  2017-12-22        49.6815  49.6815  49.6815        49.6815     116   
299  2017-12-26        50.2702  50.2702  50.2702        50.2702    9000   

         Time  
0    19:59:00  
1    18:57:00  
2    19:48:00  
3    19:48:00  
4    19:59:00  
..        ...  
295  17:29:00  
296  19:44:00  
297  17:29:00  
298  18:59:00  
299  17:41:00  

[300 rows x 7 columns]

In [15]:
dfA1.iloc[0,1]

14.8242

In [16]:
fig = go.Figure(data=go.Scatter(x=dfAA1daybyday["Date"], y=dfAA1daybyday["Closing Price"], mode='lines'))

## get the day by day data for a particular stock
fig.update_layout(
    title='Closing Price',
    xaxis=dict(title='Time'),
    yaxis=dict(title='Price')
)

# Show the plot
fig.show()


### Q 1(c)

In [17]:
def getcstick(df):
    df=df.head(10)
    groups=df.groupby("Date")
    #data1=pd.DataFrame(columns=['Date','High','Opening Price','High','Low','Closing Price','Volume'])
    data1={'Date':[],'Opening Price':[],'High':[],'Low':[],"Closing Price":[],'Volume':[]}
    df=pd.DataFrame(data1)
    for name,group in groups:
        data={'Date':name,'Opening Price':group.iloc[0,1],'High':group["High"].max(),'Low':group["Low"].min(),"Closing Price":group.iloc[group.shape[0]-1,4],'Volume':group["Volume"].sum()}
        print(data1)
        # data.append(group.iloc[0,1])
        # data.append(group["High"].max())
        # data.append(group["Low"].min())
        # data.append(group.iloc[group.shape[0]-1,4])
        # data.append(group["Volume"].sum())
        #d1=pd.DataFrame(data)
        # print(data)
        # print(len(data))
        #data=pd.DataFrame(data)
        #pd.concat([data1,data],ignore_index=True)
        df=df.append(data,ignore_index=True)
        # data1.append(data)   
    #result_df = pd.DataFrame(data1)
    df=pd.DataFrame(data1)
    return data1

In [18]:
df_cstick=getcstick(dfA1)
df_cstick

{'Date': [], 'Opening Price': [], 'High': [], 'Low': [], 'Closing Price': [], 'Volume': []}


C:\Users\kumar\AppData\Local\Temp\ipykernel_18140\97090256.py:20: FutureWarning:

The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.



{'Date': [],
 'Opening Price': [],
 'High': [],
 'Low': [],
 'Closing Price': [],
 'Volume': []}

In [19]:
import plotly.graph_objs as go
# Group by 'Date' and calculate the desired values for each date
dfA1final = dfA1.groupby('Date').agg({
    'Opening Price': 'first',    # First opening price
    'High': 'max',         # Maximum high
    'Low': 'min',       # Minimum low
    'Closing Price': 'last',      # Last closing price
    'Volume' : 'sum'    # Volume sum
}).reset_index()

# Create the daily candlestick chart for the first 200 days
first_200_days_data = dfA1final.head(200).copy()
fig = go.Figure()

# Candlestick trace
trace = go.Candlestick(
    x=first_200_days_data['Date'],
    open=first_200_days_data['Opening Price'],
    high=first_200_days_data['High'],
    low=first_200_days_data['Low'],
    close=first_200_days_data['Closing Price'],
    name='Candlestick',
    showlegend=True,
)
fig.add_trace(trace)

# Volume trace on the secondary y-axis
volume_trace = go.Bar(
    x=first_200_days_data['Date'],
    y=first_200_days_data['Volume'],
    yaxis='y2',
    name='Volume',
    showlegend=True,
)
fig.add_trace(volume_trace)

# Configure layout
fig.update_layout(
    title='Daily Candlestick Chart with Volume for A1',
    yaxis_title='Price',
    yaxis2=dict(
        title='Volume',
        overlaying='y',
        side='right',
    ),
    xaxis=dict(
        title='Date',
    ),
)

# Show the chart
fig.show()

##### Observations:

#### Large Jumps: The candlestick chart shows significant price jumps at about 3 points along the timeline. These jumps can indicate sudden and substantial price movements in the underlying asset.
#### Missing Data: There are many points in the candlestick chart where data is missing. These gaps in the chart can make it challenging to analyze the price action during those periods and may require further investigation to understand the reasons for the missing data.

In [20]:
import plotly.graph_objs as go


# Group by 'Date' and calculate the desired values for each date
dfAA1final = dfAA1.groupby('Date').agg({
    'Opening Price': 'first',    # First opening price
    'High': 'max',         # Maximum high
    'Low': 'min',       # Minimum low
    'Closing Price': 'last',      # Last closing price
    'Volume' : 'sum'    # Volume sum
}).reset_index()

# Create the daily candlestick chart for the first 200 days
first_200_days_data = dfAA1final.head(200).copy()
fig = go.Figure()

# Candlestick trace
trace = go.Candlestick(
    x=first_200_days_data['Date'],
    open=first_200_days_data['Opening Price'],
    high=first_200_days_data['High'],
    low=first_200_days_data['Low'],
    close=first_200_days_data['Closing Price'],
    name='Candlestick',
    showlegend=True,
)
fig.add_trace(trace)

# Volume trace on the secondary y-axis
volume_trace = go.Bar(
    x=first_200_days_data['Date'],
    y=first_200_days_data['Volume'],
    yaxis='y2',
    name='Volume',
    showlegend=True,
)
fig.add_trace(volume_trace)

# Configure layout
fig.update_layout(
    title='Daily Candlestick Chart with Volume for AA1',
    yaxis_title='Price',
    yaxis2=dict(
        title='Volume',
        overlaying='y',
        side='right',
    ),
    xaxis=dict(
        title='Date',
    ),
)

# Show the chart
fig.show()

##### Observations:

#### Large Jumps: The candlestick chart shows significant price jumps at about 2 points along the timeline. These jumps can indicate sudden and substantial price movements in the underlying asset.
#### Missing Data: There are many points in the candlestick chart where data is missing. These gaps in the chart can make it challenging to analyze the price action during those periods and may require further investigation to understand the reasons for the missing data.

### Q 2

##### Min-Max Scaling: MinMax scaling transforms the data into a specific range, typically [0, 1] or [-1, 1], ensuring that all features have the same scale. This is particularly useful when you want to preserve the relationships between the original data points and ensure that the scaled values fall within a predefined range. It's beneficial when you want to compare or visualize different features on a common scale. For stock price prediction, MinMax scaling can be helpful if you have features with vastly different value ranges, as it prevents one feature from dominating the others in the modeling process

In [21]:
## normalization for A1 stock using z score normalization
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,MinMaxScaler
# Define the columns to normalize
normalize_col=["Opening Price","High","Low","Closing Price","Volume"]
label=["Closing Price"]
# Extract the selected columns for the features (X) and the target (y)
stock_X=dfA1final[normalize_col]
stock_y=dfA1final[label]
# Split the data into training, testing, and validation sets
stock_x_train,stock_x_test,stock_y_train,stock_y_test=train_test_split(stock_X,stock_y,test_size=0.15,random_state=42,shuffle=False)
stock_x_train,stock_x_val,stock_y_train,stock_y_val=train_test_split(stock_x_train,stock_y_train,test_size=0.15,random_state=42,shuffle=False)

# Initialize the Min-Max scaler for feature normalization
scaler=MinMaxScaler()
# Fit the Min-Max scaler to the training data
scaler.fit(stock_x_train)

# Transform the training, testing, and validation data using the Min-Max scaler
stock_x_train_scaled=scaler.transform(stock_x_train)
stock_x_test_scaled=scaler.transform(stock_x_test)
stock_x_val_scaled=scaler.transform(stock_x_val)
# Initialize the scaler for target variable normalization
scaler1=MinMaxScaler()
# Fit the scaler to the training target variable
scaler1.fit(stock_y_train)

# Transform the training, testing, and validation target variables using the Standard scaler
stock_y_train_scaled=scaler1.transform(stock_y_train)
stock_y_test_scaled=scaler1.transform(stock_y_test)
stock_y_val_scaled=scaler1.transform(stock_y_val)
#print(stock_y_train_scaled)
# Create DataFrames for the scaled feature sets
df_x_train=pd.DataFrame(stock_x_train_scaled,columns=[normalize_col])
df_x_test=pd.DataFrame(stock_x_test_scaled,columns=[normalize_col])
df_x_val=pd.DataFrame(stock_x_val_scaled,columns=[normalize_col])
# Create DataFrames for the scaled target variable sets
df_y_train=pd.DataFrame(stock_y_train_scaled,columns=[label])
df_y_test=pd.DataFrame(stock_y_test_scaled,columns=[label])
df_y_val=pd.DataFrame(stock_y_val_scaled,columns=[label])


In [22]:
## print the normalized data frame
print(df_x_train,df_y_train)

     Opening Price      High       Low Closing Price    Volume
0         0.138844  0.129841  0.139323      0.128558  0.088474
1         0.134759  0.127469  0.133487      0.125514  0.102803
2         0.126972  0.123226  0.134395      0.125261  0.104834
3         0.129907  0.120605  0.128819      0.118921  0.083125
4         0.120715  0.114490  0.128041      0.118374  0.067636
...            ...       ...       ...           ...       ...
3114      0.956629  0.929465  0.964874      0.937853  0.037523
3115      0.950434  0.927511  0.963453      0.939441  0.041601
3116      0.950235  0.939039  0.968733      0.954333  0.028997
3117      0.960427  0.978897  0.968528      1.000000  0.113423
3118      1.000000  1.000000  1.000000      0.998014  0.215609

[3119 rows x 5 columns]      Closing Price
0         0.128558
1         0.125514
2         0.125261
3         0.118921
4         0.118374
...            ...
3114      0.937853
3115      0.939441
3116      0.954333
3117      1.000000
3118      

In [23]:
## print the normalized test data frame
df_x_test

Opening Price      High       Low Closing Price    Volume
0        1.237864  1.206211  1.233842      1.209527  0.058004
1        1.197901  1.166944  1.177398      1.143695  0.071562
2        1.185259  1.177109  1.202463      1.201020  0.045190
3        1.194638  1.188870  1.211575      1.207298  0.037131
4        1.226445  1.213586  1.245130      1.238087  0.035099
..            ...       ...       ...           ...       ...
643      2.539643  2.535260  2.575463      2.553135  0.078376
644      2.565171  2.520244  2.549318      2.531486  0.051549
645      2.526775  2.554331  2.490073      2.428396  0.074041
646      2.419267  2.437871  2.385496      2.472724  0.064762
647      2.505191  2.541752  2.531608      2.587773  0.055639

[648 rows x 5 columns]

In [24]:
stock_x_train_scaled.shape[0]

3119

### Q 3(a)

#### Here I made a decision to perform inter day trading on the stocks  

In [25]:
## modify data for stocks for per day trading
dfA1dbd= dfA1.groupby('Date').agg({
    'Opening Price': 'first',    # First opening price
    'High': 'max',         # Maximum high
    'Low': 'min',       # Minimum low
    'Closing Price': 'last',      # Last closing price
    'Volume' : 'sum'    # Volume sum
}).reset_index()
dfA1dbd

Date  Opening Price      High       Low  Closing Price   Volume
0     2005-01-03        14.8242   14.8734   14.4551        14.4551  3218758
1     2005-01-04        14.6274   14.7565   14.1783        14.3075  3655769
2     2005-01-05        14.2522   14.5474   14.2214        14.2952  3717698
3     2005-01-06        14.3936   14.4182   13.9569        13.9877  3055615
4     2005-01-07        13.9507   14.1168   13.9200        13.9612  2583226
...          ...            ...       ...       ...            ...      ...
4313  2022-02-18       130.5000  133.4300  130.0000       132.0500  2910789
4314  2022-02-22       131.7300  132.6899  128.7600       131.0000  2092609
4315  2022-02-23       129.8800  134.3700  125.9500       126.0000  2778577
4316  2022-02-24       124.7000  128.6300  120.9900       128.1500  2495574
4317  2022-02-25       128.8400  133.7500  127.9200       133.7300  2217367

[4318 rows x 6 columns]

In [26]:
## modify data for stocks for per day trading
dfAA1dbd= dfAA1.groupby('Date').agg({
    'Opening Price': 'first',    # First opening price
    'High': 'max',         # Maximum high
    'Low': 'min',       # Minimum low
    'Closing Price': 'last',      # Last closing price
    'Volume' : 'sum'    # Volume sum
}).reset_index()
dfAA1dbd

Date  Opening Price     High      Low  Closing Price    Volume
0     2016-10-18        26.5376  26.5972  25.4741        26.2394  12359405
1     2016-10-19        26.4382  27.2631  26.1897        26.8854   9478623
2     2016-10-20        26.8755  27.0097  26.4531        26.6171   7556726
3     2016-10-21        26.5376  26.8954  26.1698        26.7662   6026468
4     2016-10-24        26.8357  27.2532  26.5972        26.7562   5459936
...          ...            ...      ...      ...            ...       ...
1344  2022-02-18        77.2200  78.4000  75.5500        78.1500   4638821
1345  2022-02-22        79.7500  81.8900  73.1100        74.4000   7333403
1346  2022-02-23        74.5000  76.1300  73.1101        75.5500   5407063
1347  2022-02-24        77.9600  78.0000  69.8500        73.0000   9745029
1348  2022-02-25        72.6100  78.3800  71.0700        77.7900   5961794

[1349 rows x 6 columns]

#### Q 3(b)
#### I have assummed the Commission's plus the buy ask spread to be about 2 percent

#### Q 3(c)
#### I decided to fit my model only on one stock for the purpose of ease of training and so that the model can capture information from one stock accurately

### Q (4)

In [27]:
### LSTM model
import torch
import torch.nn as nn
# Define the LSTM model class
class Lstm(nn.Module):
    def __init__(self,input_size,hidden_size,nlayers,output_size):
        super(Lstm,self).__init__()
        # Initialize the number of LSTM layers and define the LSTM module
        self.nlayers=nlayers
        self.lstm=nn.LSTM(input_size,hidden_size,nlayers,batch_first=True)
        # Define the fully connected (linear) layer for output
        self.fc=nn.Linear(hidden_size,output_size)


    def forward(self,x):
        ##initialize hidden state 0
        # h0=torch.zeros(self.nlayers,x.size(0),self.hidden_size)
        # c0=torch.zeros(self.nlayers,x.size(0),self.hidden_size)
        # Pass the input sequence through the LSTM layers
        out, _=self.lstm(x)
        # Pass the LSTM output through the fully connected layer for final prediction
        out=self.fc(out)
        #print(out.size())
        return out


In [28]:
dfA1dbd.shape[0]

4318

### Q (5)

In [29]:
dfA1dbd.iloc[1:2,4:5]

Closing Price
1        14.3075

In [30]:
df1=dfA1.head(5)
df1.iloc[3-1-1:3-1,0:df1.shape[1]-2].values.tolist()
#torch.tensor(df1.iloc[3-1-1:3-1,1:-1].values.tolist())

[[14.8242, 14.8242, 14.8119, 14.8119, 9646.0]]

In [31]:
df_y_train.iloc[31:32,:]

Closing Price
31      0.134011

In [32]:
##### Custom Sequencer which provides input to the dataoader here 

In [33]:
# Import necessary libraries
from torch.utils.data import Dataset
# Define a custom dataset class
class CustomDataset(Dataset):
    def __init__(self, data, labels, length):
        # Initialize the dataset with data, labels, and sequence length
        self.data = data
        self.labels = labels
        self.length = length

    def __len__(self):
        # Define the length of the dataset
        return len(self.data) - self.length - 1

    def __getitem__(self, index):
        # Initialize empty lists for sequence data and labels
        seq_data = []
        out = []
        # Extract a sequence of data and corresponding labels
        seq_data.append(self.data.iloc[index:index+30, :].values.tolist())
        out.append(self.labels.iloc[index+1:index+31, :].values.tolist())
        # Flatten the output list and convert it to a NumPy array
        out = np.array(out).flatten()
        # Convert the sequence data and output to PyTorch tensors
        seq_data = torch.Tensor(seq_data)
        out = torch.Tensor(out)
        return seq_data, out


In [34]:
dfA1

Opening Price      High       Low  Closing Price  Volume        Date  \
0              14.8242   14.8242   14.8181        14.8242  173492  2005-01-03   
1              14.8242   14.8242   14.8119        14.8119    9646  2005-01-03   
2              14.8119   14.8242   14.8119        14.8181   34810  2005-01-03   
3              14.8181   14.8242   14.8181        14.8242   11464  2005-01-03   
4              14.8242   14.8242   14.8181        14.8181   12302  2005-01-03   
...                ...       ...       ...            ...     ...         ...   
1680045       133.7300  133.7300  133.7300       133.7300  354729  2022-02-25   
1680046       133.7300  133.7300  133.7300       133.7300    6777  2022-02-25   
1680047       133.7300  133.7300  133.7300       133.7300       0  2022-02-25   
1680048       133.7300  133.7300  133.7300       133.7300       0  2022-02-25   
1680049       133.7300  133.7300  133.7300       133.7300       0  2022-02-25   

             Time  
0        09:30:00  
1        09:31:00  
2        09:32:00  
3        09:33:00  
4        09:34:00  
...           ...  
1680045  16:02:00  
1680046  16:03:00  
1680047  16:10:00  
1680048  18:30:00  
1680049  19:00:00  

[1680050 rows x 7 columns]

In [35]:
# Retrieve a sequence of data and labels from the custom dataset at index 0
a, b = CustomDataset(df_x_train, df_y_train, 30).__getitem__(0)
# Print the size (shape) of the variable 'b'
print(b.size())

torch.Size([30])


In [36]:
# Create a custom dataset for training using df_x_train and df_y_train
custom_dataset_train = CustomDataset(df_x_train, df_y_train, 30)
# Create a custom dataset for testing using df_x_test and df_y_test
custom_dataset_test = CustomDataset(df_x_test, df_y_test, 30)
# Create a custom dataset for validation using df_x_val and df_y_val
custom_dataset_val = CustomDataset(df_x_val, df_y_val, 30)

In [37]:
print(CustomDataset(df_x_train,df_y_train,30).__getitem__(12))

(tensor([[[0.1128, 0.1096, 0.1205, 0.1150, 0.1610],
         [0.1152, 0.1059, 0.1162, 0.1070, 0.1323],
         [0.1097, 0.1009, 0.1125, 0.1032, 0.1435],
         [0.1058, 0.1026, 0.1127, 0.1070, 0.0790],
         [0.1109, 0.1035, 0.1175, 0.1095, 0.0636],
         [0.1109, 0.1055, 0.1160, 0.1095, 0.0683],
         [0.1108, 0.1033, 0.1135, 0.1046, 0.0704],
         [0.1113, 0.1058, 0.1186, 0.1100, 0.0673],
         [0.1122, 0.1095, 0.1200, 0.1161, 0.0718],
         [0.1188, 0.1122, 0.1244, 0.1184, 0.0594],
         [0.1187, 0.1112, 0.1179, 0.1109, 0.0676],
         [0.1127, 0.1086, 0.1201, 0.1152, 0.0840],
         [0.1248, 0.1177, 0.1289, 0.1237, 0.1040],
         [0.1265, 0.1275, 0.1335, 0.1317, 0.1223],
         [0.1349, 0.1250, 0.1323, 0.1234, 0.0636],
         [0.1259, 0.1190, 0.1327, 0.1227, 0.0430],
         [0.1245, 0.1253, 0.1304, 0.1311, 0.0702],
         [0.1317, 0.1280, 0.1296, 0.1286, 0.1119],
         [0.1280, 0.1342, 0.1361, 0.1398, 0.3064],
         [0.1399, 0.1308, 0.14

In [38]:
from torch.utils.data import DataLoader
batch_size = 16
shuffle = False  # Set shuffle to false because it is a sequential time series data shuffle true will cause leaking into predictions
# Create a DataLoader for the training dataset
dataloader_train = DataLoader(custom_dataset_train, batch_size=batch_size, shuffle=shuffle)
# Create a DataLoader for the testing dataset
dataloader_test = DataLoader(custom_dataset_test, batch_size=batch_size, shuffle=shuffle)
# Create a DataLoader for the validation dataset
dataloader_val = DataLoader(custom_dataset_val, batch_size=batch_size, shuffle=shuffle)

In [39]:
print(dataloader_train)

In [40]:
from transformers import AdamW

### Q(6)

In [44]:
## Define a lstm model based on the previous custom defined lstm model with input size as tensors of dimension 5 cross 1 and with a hidden layer of size 120 and 1 layer
## These values were obtained by me afte hyperparameter tuning while training the model
model=Lstm(5,150,1,1)

In [45]:
import torch.nn as nn
## define the optimizer here as AdamW with learning rate as 3*10^(-4)
optimizer  = AdamW(model.parameters(),lr = 3e-4)
## define the loss function as MSE Loss
lossfn=nn.MSELoss()

##### Training

In [46]:
epochs=10 # Set the number of epochs
import random
a=random.randint(0,30)
losses=[] # Initialize a list to store losses
for no in range(epochs): # Loop over the specified number of epochs
        loss=0
        netloss=0
        model.train()
        print("/n")
        print("training"+str(no+1)+"/"+str(epochs))  # Print epoch information
        for step,batch in enumerate(dataloader_train): # Iterate through batches in the training data loader
                t_data,t_labels=batch
                t_data=t_data.squeeze(1)
                t_labels=t_labels.squeeze(-1)
                t_labels=t_labels.squeeze(-1)
                optimizer.zero_grad()
                output=model(t_data).squeeze(-1)
                # if(step==1):print(output[1][a],t_labels[1][a])
                # print(output.size(),t_labels.size())
                #if(step==1):print(output.flatten(),t_data.flatten())
                loss=lossfn(output,t_labels) # Calculate and backpropagate the loss
                netloss+=loss
                loss.backward()
                torch.nn.utils.clip_grad_norm_(model.parameters(),1.0)
                optimizer.step()
        avg_loss=netloss/(len(dataloader_train)) # Calculate and print the average training loss
        print("Average Training Loss "+str(avg_loss))
        # losses.appned(avg_loss)
        # Validation phase
        print("Validation")
        model.eval()
        loss=0
        netloss=0

        # eval_acc=0
        # eval_steps=0
        # acc=0
        for step,batch in enumerate(dataloader_val): # Iterate through batches in the validation data loader
                t_data,t_labels=batch
                t_data=t_data.squeeze(1)
                t_labels=t_labels.squeeze(-1)
                t_labels=t_labels.squeeze(-1)
                optimizer.zero_grad()
                output=model(t_data).squeeze(-1)
                loss=lossfn(output,t_labels) # Calculate validation loss
                netloss+=loss
                #print("Loss"+str(loss))
                # acc=flat_accuracy(output,t_labels)
                # eval_acc+=acc
                # eval_steps+=1
                # oacc=eval_acc/eval_steps
        print("Validation Loss is "+str(netloss/(len(dataloader_val))))

/n
training1/10


Average Training Loss tensor(0.0088, grad_fn=<DivBackward0>)
Validation
Validation Loss is tensor(0.0452, grad_fn=<DivBackward0>)
/n
training2/10
Average Training Loss tensor(0.0063, grad_fn=<DivBackward0>)
Validation
Validation Loss is tensor(0.0211, grad_fn=<DivBackward0>)
/n
training3/10
Average Training Loss tensor(0.0028, grad_fn=<DivBackward0>)
Validation
Validation Loss is tensor(0.0112, grad_fn=<DivBackward0>)
/n
training4/10
Average Training Loss tensor(0.0010, grad_fn=<DivBackward0>)
Validation
Validation Loss is tensor(0.0057, grad_fn=<DivBackward0>)
/n
training5/10
Average Training Loss tensor(0.0006, grad_fn=<DivBackward0>)
Validation
Validation Loss is tensor(0.0031, grad_fn=<DivBackward0>)
/n
training6/10
Average Training Loss tensor(0.0004, grad_fn=<DivBackward0>)
Validation
Validation Loss is tensor(0.0017, grad_fn=<DivBackward0>)
/n
training7/10
Average Training Loss tensor(0.0003, grad_fn=<DivBackward0>)
Validation
Validation Loss is tensor(0.0010, grad_fn=<DivBackwa

### Q (7)

##### 1. Buy Condition:If either the short-term moving average (short_avg) is less than the long-term moving average (long_avg)or the predicted price (pred) is greater than the current price (current).Additionally, ensure that the balance allows for buying at least one share (int(self.balance/current) > 0).If these conditions are met, the decision is set to "buy."
##### 2. Sell Condition:If either the short-term moving average (short_avg) is greater than the long-term moving average (long_avg)or the predicted price (pred) is less than the current price (current).Additionally, check if there are stocks held (self.stock_held > 0).If these conditions are met, the decision is set to "sell."
##### 3. Hold Condition:If none of the above conditions are met, the default decision is set to "hold."This means that if neither a strong buy nor a strong sell signal is present, the trader holds the current position.

In [66]:
class trading_module:
    def __init__(self,balance,commission):
        # Initialize the trading module with balance and commission parameters
        self.balance = balance  
        self.stock_held = 0 
        self.commission=commission
    def buy(self,current):
        # Buy method for executing a buy trade
        if self.balance > 0:
            # Calculate the number of stocks you can buy
            num_to_buy = self.balance // current
            # Calculate the cost, including commissions
            cost = num_to_buy * current+self.commission*current
            # Update the balance and the number of stocks held
            self.balance -= cost
            self.stock_held += num_to_buy
            print(f"Bought {num_to_buy} stocks at ${current} each, balance ${self.balance}.")
        else:
            print("LB ")
    def sell(self,current):
        if self.stock_held > 0:
            # Calculate the revenue from selling stocks, minus commissions
            revenue = self.stock_held * current-self.commission*current
            self.balance += revenue
            print(f"Sold {self.stock_held} stocks at ${current} each , balance ${self.balance}.")
            self.stock_held = 0
        else:
            print("NS ")
    def hold(self):
        print("H ")
    def make_decision(self,short_avg,long_avg,pred,current):
        ## This is where I have implemeted muy Tradinig Startegy
        # 1. Buy Condition:
        # If either the short-term moving average (short_avg) is less than the long-term moving average (long_avg)
        # or the predicted price (pred) is greater than the current price (current).
        # Additionally, ensure that the balance allows for buying at least one share (int(self.balance/current) > 0).
        # If these conditions are met, the decision is set to "buy."
        if(((1.3*short_avg<long_avg) or (pred>1.1*current)) and int(self.balance/current)>0):
            decision="buy"
        # 2. Sell Condition:
        # If either the short-term moving average (short_avg) is greater than the long-term moving average (long_avg)
        # or the predicted price (pred) is less than the current price (current).
        # Additionally, check if there are stocks held (self.stock_held > 0).
        # If these conditions are met, the decision is set to "sell."
        elif(((short_avg>1.3*long_avg)or (pred<0.9*current)) and self.stock_held>0):
            decision="sell"
        # 3. Hold Condition:
        # If none of the above conditions are met, the default decision is set to "hold."
        # This means that if neither a strong buy nor a strong sell signal is present, the trader holds the current position.
        else:
            decision="hold"
        return decision
    def execute_trade(self,short_avg,long_avg,pred,current):
        decision = self.make_decision(short_avg,long_avg,pred,current)
        if decision == "buy":
            self.buy(current)
        elif decision == "sell":
            self.sell(current)
        elif decision == "hold":
            self.hold()

In [67]:
class StockTrader:
    def __init__(self, initial_balance,df,commission):
        # Initialize the StockTrader with an initial balance, historical stock prices, and trading commission.
        self.initial_balance = initial_balance
        self.balance = initial_balance
        self.stock_prices = np.array(df).flatten()
        self.trading_module = trading_module(balance=initial_balance,commission=commission)
        self.net_profit = 0

    def add_stock_prices(self, df):
        self.stock_prices = np.array(df).flatten()

    def trade_stock(self):
        # print(self.stock_prices)
        days=0
        total_price=0
        for days,stock_price in enumerate(self.stock_prices):
            days+=1
            # print(str(days)+" "+"index")
            total_price+=self.stock_prices[days-1]
            
            long_avg=(1/days)*total_price
            short_avg=0
            if(days<=10): short_avg=long_avg
            else :
                for i in range(10):
                    short_avg+=self.stock_prices[days-i-1]
                    short_avg=(short_avg/10)
            # self.trading_module.stock_price = stock_price
            a, b = CustomDataset(df_x_test, df_y_test, 30).__getitem__(days-1)
            # Make a prediction for the last element in the sequence
            pred = model(a)[-1][-1].flatten().item()
            current=self.stock_prices[days-1]
            decision = self.trading_module.execute_trade(short_avg,long_avg,pred,current)
        
            # if decision == "buy":
            #     self.trading_module.buy()
            # elif decision == "sell":
            #     self.trading_module.sell()
        
        # Calculate net profit at the end of trading
        self.net_profit = self.balance - self.initial_balance

    def get_net_profit(self):
        return self.net_profit

### Q (8)

##### Test and plot the predicted stock price values in comparison to the actual stock prices.

In [68]:
# Initialize variables to store loss
loss = 0
net_loss = 0
# Set the model to evaluation mode
model.eval()
# Disable gradient computation during evaluation
with torch.no_grad():
    # Iterate through batches in the test data loader
    for step, batch in enumerate(dataloader_test):
        loss = 0
        t_data, t_labels = batch
        t_data = t_data.squeeze(1)
        t_labels = t_labels.squeeze(-1)
        t_labels = t_labels.squeeze(-1)
        optimizer.zero_grad()
        output = model(t_data).squeeze(-1)
        # Calculate the test loss
        loss = lossfn(output, t_labels)
        net_loss += loss
    # Print the test loss
    print("Test loss is {}".format(net_loss / (len(dataloader_test))))

Test loss is 0.016904564574360847


In [69]:
import random
m=random.randint(0,200)
a,b=CustomDataset(df_x_test,df_y_test,30).__getitem__(m)
print(b[-1],model(a)[-1][-1].flatten())
# print(model(a).item()*np.sqrt(scaler1.var_)+scaler1.mean_,b.item()*np.sqrt(scaler1.var_)+scaler1.mean_)

tensor(1.4019) tensor([1.4108], grad_fn=<SelectBackward0>)


In [70]:
# Initialize empty lists to store predictions and actual values
pred = []
actual = []
# Iterate through the training data
for i in range(len(df_y_train) - 1):
    # Retrieve a sequence of data and labels from the custom dataset at index 'i'
    a, b = CustomDataset(df_x_train, df_y_train, 30).__getitem__(i)
    # Make a prediction for the last element in the sequence
    prediction = model(a)[-1][-1].flatten().item()
    pred.append(prediction)
    # Get the actual value for the last element in the sequence
    actual_value = b[-1].item()
    actual.append(actual_value)
# Print the length of the 'pred' list (number of predictions made)
print(len(pred))


3118


In [71]:
# pred = []
# actual = []
# Iterate through the validation data
for i in range(len(df_y_val) - 1):
    # Retrieve a sequence of data and labels from the custom dataset at index 'i'
    a, b = CustomDataset(df_x_val, df_y_val, 30).__getitem__(i)
    # Make a prediction for the last element in the sequence
    prediction = model(a)[-1][-1].flatten().item()
    pred.append(prediction)
    # Get the actual value for the last element in the sequence
    actual_value = b[-1].item()
    actual.append(actual_value)
# Print the length of the 'pred' list (number of predictions made)
print(len(pred))

3668


In [72]:
# pred = []
# actual = []
# Iterate through the test data
for i in range(len(df_y_test) - 1):
    # Retrieve a sequence of data and labels from the custom dataset at index 'i'
    a, b = CustomDataset(df_x_test, df_y_test, 30).__getitem__(i)
    # Make a prediction for the last element in the sequence
    prediction = model(a)[-1][-1].flatten().item()
    pred.append(prediction)
    # Get the actual value for the last element in the sequence
    actual_value = b[-1].item()
    actual.append(actual_value)
# Convert the 'pred' list to a NumPy array and flatten it
pred = np.array(pred).flatten()
# Print the length of the 'pred' array (number of predictions made)
print(len(pred))

4315


In [73]:
import plotly.graph_objs as go
import plotly.offline as pyo

In [74]:
pred=pred*(scaler1.data_max_-scaler1.data_min_)+scaler1.data_min_
actual=actual*(scaler1.data_max_-scaler1.data_min_)+scaler1.data_min_

In [75]:
## Print the values for predicted values and actual values of the test dataset
predicted = go.Scatter(y=pred[3667:4313], mode='lines', name='Predicted')
# Create a Scatter plot for actual values
actualval = go.Scatter(y=actual[3667:4313], mode='lines', name='Actual')
layout = go.Layout(title='Actual vs. Predicted Stock Prices', xaxis=dict(title='Days'), yaxis=dict(title='Price'))
# Create a Figure containing both predicted and actual data
fig = go.Figure(data=[predicted], layout=layout)
fig.add_trace(actualval)
# Use Plotly's iplot function to display the interactive plot
pyo.iplot(fig)

In [76]:
print(len(pred[3667:4313]))

646


### Q 8(a)

In [77]:
## Print the values for predicted values and actual values of the test dataset
percenatge_error=((actual[3667:4313]-pred[3667:4313])*100)/actual[3667:4313]
error=go.Scatter(y=percenatge_error,mode= 'lines', name='Percentage Error')
fig = go.Figure(data=[error])
#fig.add_trace(actualval)
pyo.iplot(fig)

#### Percentage Error between the predicted values i.e. ((actual - predicted )/actual) increases as time passes or the number of days of testing increases.

#### Q 8(b)

In [78]:
## convert the test values int0 a list for feeding it into the dataset
l1=df_y_test.values.tolist()*(scaler1.data_max_-scaler1.data_min_)+scaler1.data_min_

In [79]:
St1=StockTrader(10000,l1,0)
St1.add_stock_prices(l1)
St1.trade_stock()

H 
H 
H 
H 
H 
H 
H 
H 
H 
H 
Bought 143.0 stocks at $69.8307 each, balance $14.209900000001653.
Sold 143.0 stocks at $70.4889 each , balance $10094.122600000002.
Bought 144.0 stocks at $69.811 each, balance $41.33860000000095.
Sold 144.0 stocks at $70.931 each , balance $10255.402600000001.
Bought 145.0 stocks at $70.4791 each, balance $35.933100000000195.
Sold 145.0 stocks at $68.3276 each , balance $9943.4351.
Bought 145.0 stocks at $68.416 each, balance $23.115100000000893.
Sold 145.0 stocks at $68.2588 each , balance $9920.6411.
Bought 144.0 stocks at $68.7107 each, balance $26.30030000000079.
Sold 144.0 stocks at $69.4573 each , balance $10028.151500000002.
Bought 143.0 stocks at $69.8601 each, balance $38.15720000000147.
Sold 143.0 stocks at $69.1724 each , balance $9929.8104.
Bought 142.0 stocks at $69.6145 each, balance $44.551399999998466.
Sold 142.0 stocks at $72.2278 each , balance $10300.899.
Bought 141.0 stocks at $72.9744 each, balance $11.508599999999205.
Sold 141.0 sto

In [80]:
net_profit = St1.trading_module.balance-St1.balance
print(f"Net Profit: ${net_profit}")

Net Profit: $7358.625000000033


#### The net profit after considering 0% commission is about 7.3k Dollars 

In [89]:
St2=StockTrader(10000,l1,0.02)
St2.add_stock_prices(l1)
St2.trade_stock()

H 
H 
H 
H 
H 
H 
H 
H 
H 
H 
Bought 143.0 stocks at $69.8307 each, balance $12.813286000002336.
Sold 143.0 stocks at $70.4889 each , balance $10091.316208000004.
Bought 144.0 stocks at $69.811 each, balance $37.13598800000182.
Sold 144.0 stocks at $70.931 each , balance $10249.781368000002.
Bought 145.0 stocks at $70.4791 each, balance $28.90228600000046.
Sold 145.0 stocks at $68.3276 each , balance $9935.037734000001.
Bought 145.0 stocks at $68.416 each, balance $13.34941400000207.
Sold 145.0 stocks at $68.2588 each , balance $9909.510238000003.
Bought 144.0 stocks at $68.7107 each, balance $13.795224000003145.
Sold 144.0 stocks at $69.4573 each , balance $10014.257278000005.
Bought 143.0 stocks at $69.8601 each, balance $22.86577600000419.
Sold 143.0 stocks at $69.1724 each , balance $9913.135528000003.
Bought 142.0 stocks at $69.6145 each, balance $26.484238000000914.
Sold 142.0 stocks at $72.2278 each , balance $10281.387282000002.
Bought 140.0 stocks at $72.9744 each, balance $63

In [90]:
net_profit = St2.trading_module.balance-St2.balance
print(f"Net Profit: ${net_profit}")

Net Profit: $2038.0266740000188


#### The net profit after considering 0% commission is about 2.7k Dollars i.e. a net overall profit of about 3k Dollars in a balance of 10k Dollars

#### Q 8(c)

#### Simple Buy and Hold

In [91]:
balance=10000
nstocks_inti=balance/l1.flatten()[0]
final_balance=nstocks_inti*l1.flatten()[-1]

In [92]:
profit_2=(final_balance-balance)
print("Buy and Sell Profit "+str(profit_2)+" Trading Startegy profit "+str(net_profit))

Buy and Sell Profit 9994.497887368165 Trading Startegy profit 2038.0266740000188


#### Here I see that the buy and Hold strategy gives me a better profit as comapred to my trading startegy over 646 days i.e. about 2 years

#### Here is the link to the video assignnment https://drive.google.com/drive/folders/1ny8wgOf3RqqZO-2G0sMCSxUFTzvbLT_Z?usp=drive_link